envoy-control-core/src/main/kotlin/pl/allegro/tech/servicemesh/envoycontrol/snapshot/listeners/filters/RBACFilterFactory.kt

```diff
@@ -23,23 +24,25 @@ class RBACFilterFactory(
         private val logger by logger()
     }
 
-    fun getRules(serviceName: String, incomingPermissions: Incoming): RBAC {
-        val clientToPolicy = mutableMapOf<String, Policy>()
+    private fun getRules(serviceName: String, incomingPermissions: Incoming): RBAC {
+        val clientToPolicyBuilder = mutableMapOf<String, Policy.Builder>()
         incomingPermissions.endpoints.forEach { incomingEndpoint ->
             if (incomingEndpoint.clients.isEmpty()) {
-                logger.error("An incoming endpoint definition for $serviceName does not have any clients defined." +
-                        "It means that noone will be able to contact that endpoint.")
+                logger.warn("An incoming endpoint definition for $serviceName does not have any clients defined." +
+                        "It means that no one will be able to contact that endpoint.")
                 return@forEach
             }
 
-            val policy = Policy.newBuilder()
-            val clientName = incomingEndpoint.clients.joinToString(",")
+            val clients = resolveClients(incomingEndpoint, incomingPermissions.roles)
+            val policyName = clients.joinToString(",")
+
+            val policy: Policy.Builder = clientToPolicyBuilder.getOrDefault(policyName, Policy.newBuilder())
```

Comment:
If we change it to:
```kotlin
            val policy: Policy.Builder = clientToPolicyBuilder.computeIfAbsent(policyName) {
                Policy.newBuilder().addAllPrincipals(
                    clients.map(this::mapClientToPrincipal)
                )
            }
```

we can remove:

```kotlin
            val principals = clients.map(this::mapClientToPrincipal)
            policy.addAllPrincipals(principals - policy.principalsList)
```

and this way we will avoid mapping client to principal multiple times, computing principalsList difference, etc.